### DISCUSSION TITLE
hybrid property which is a tuple and is selectable as a bundle?

### DISCUSSION THREAD

**Discussion by myrix** [sqlalchemy/sqlalchemy#10769] [Usage Questions] [answered]
Using SQLAlchemy 1.4.50, trying to make a tuple hybrid property selectable as a tuple as if it were a bundle. Is it in any way possible?

The best I could achieve is to make it selectable as a list of its constituents by customizing Tuple:
```python
class Selectable_Tuple(Tuple):

    inherit_cache = True

    @property
    def _select_iterable(self):
        return super(Tuple, self)._select_iterable

class Id_Comparator(Comparator):

    def __init__(self, client_id, object_id):

        self.client_id = client_id
        self.object_id = object_id

        super().__init__(Selectable_Tuple(client_id, object_id))

    def operate(self, op, other, **kwargs):

        if other is None:
            return (
                and_(
                    op(self.client_id, None, **kwargs),
                    op(self.object_id, None, **kwargs)))

        if isinstance(other, (list, tuple)):
            return (
                and_(
                    op(self.client_id, other[0], **kwargs),
                    op(self.object_id, other[1], **kwargs)))

        return op(self.__clause_element__(), other, **kwargs))

class IdValue(Base):

    client_id = Column(BigInteger, primary_key = True)
    object_id = Column(BigInteger, primary_key = True)
    value = Column(UnicodeText)
    metadata = Column(JSONB)

    @hybrid_property
    def id(self):
        return (self.client_id, self.object_id)

    @id.setter
    def id(self, id):
        self.client_id = id[0]
        self.object_id = id[1]

    @id.comparator
    def id(cls):
        return Id_Comparator(cls.client_id, cls.object_id))
```

So right now e.g. `DBSession.query(IdValue.id, IdValue.metadata).filter(IdValue.content.in_(...))` returns 3-tuple rows of the form (client_id, object_id, metadata).

Is there a way to make the hybrid `id` property a proper bundle, so that the same query would return nested tuple rows of the form ((client_id, object_id), metadata)?

I tried making a custom class by inheriting both from Tuple and Bundle like
```python
class BundleTuple(Tuple, Bundle):
    def __init__(self, name, *args, **kwargs):
        Tuple.__init__(self, *args, **kwargs)
        Bundle.__init__(self, name, *args, **kwargs)
```
but besides me being not sure if I actually manage initialization of everything inherited right, there's a conflict, Tuple initializes its clauses like `self.clauses = [...]`, while Bundle has `self.clauses` as a property, which leads to `AttributeError: can't set attribute` during initialization.

Can't use composite column type as I would like to retain this instance-level behaviour,
```python
@hybrid_property
def id(self):
    return (self.client_id, self.object_id)
```
when I can get instance ids as just Python 2-tuples, and hopefully select them as just 2-tuples too.

I guess then something like hybrid composite bundle might do the job, but I don't know SQLAlchemy internals sufficiently to even understand if it's possible at all or not.

Any pointers on how the full required behaviour can be achieved would be much appreciated.

**Comment by zzzeek** [comment]
this seems complicated, why not use `@hybrid_property.expression` and return an actual `Bundle` directly?    I'd avoid dealing with `@comparator` if at all possible, things are extremely complicated at that level.

**Reply by myrix** [reply]
Mostly for convenience, there are other hybrid properties of the same composite 2-tuple structure using the same comparator, including nullable foreign keys, and we use comparator to be able to use e.g. `Entity.field_id == None` and `Entry.link_id != None` in `.filter(...)` invocations of the queries.

We started with `@hybrid_property.expression` which just returned a 2-tuple via `tuple_(...)`, and it was good for literal comparisons like `Entity.field_id == (66, 25)` and foreign key conditions like `Entity.field_id == Field.id`, but when we added null-conditions as above, we had to use a comparator.

Oh, I just realized, if we could make a custom `Bundle` with overloaded comparisons with 2-tuples, maybe it would work, I would look in the docs.

If it's too compicated to actually do in a more or less straightforward way with simple modifications to already existing SQLAlchemy stuff, it would be no problem, we could live with what we have now, thanks for considering our problem!

**Reply by zzzeek** [reply]
OK sounds like you are into the code then, `@expression` is a simpler version of `@comparator` so on the Comparator you'd want to implement the `__clause_element__()` method to return your Bundle class, that's where the SQL expression comes from.

**Comment by myrix** [comment, marked as answer]
Ended up with a clunky but simple solution, for any SQL ops we are presenting a tuple, otherwise we are a bundle:

```python
class Id_Comparator(Comparator):

    def __init__(self, client_id, object_id):

        self.client_id = client_id
        self.object_id = object_id

        self.tuple = tuple_(client_id, object_id)

        super().__init__(Bundle('composite_id_bundle', client_id, object_id))

    def operate(self, op, other, **kwargs):

        if other is None:
            return (
                and_(
                    op(self.client_id, None, **kwargs),
                    op(self.object_id, None, **kwargs)))

        comparator = getattr(other, 'comparator', None)

        if (comparator is not None and
            isinstance(comparator, Id_Comparator)):

            other = other.tuple

        return op(self.tuple, other, **kwargs))
```

`Bundle` objects goes to `Comparator`'s `__init__()` through `super().__init__()` and is returned through the standard `Comparator.__clause_element__()`.
